### PR TITLE
Proposed changes for PR #1644

### PR DESCRIPTION
**Pull request by v-dobrev**
For now this has only header changes. After discussion, and finalizing the interface, we'll proceed with implementation.

Edit: this PR is now ready for review.

### CODE REVIEW

File: tests/unit/general/test_umpire_mem.cpp
Comment (artv3) [inline tests/unit/general/test_umpire_mem.cpp:177]: Here, an application could make an alias for ``MemoryType::DEVICE_UMPIRE_2``

File: general/device.hpp
Comment (tomstitt) [inline general/device.hpp:277]: I like the these `Set{Host,Device}MemoryType` methods!

File: general/mem_manager.hpp
Comment (tomstitt) [reply general/mem_manager.hpp:44]: Yea

File: mesh/mesh.hpp
Comment (tomstitt) [inline mesh/mesh.hpp:782]: We'll need an option to pass in a `MemoryType`, did you have something else in mind?

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [inline tests/unit/general/test_umpire_mem.cpp:71]: This is definitely cleaner, nice!

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [inline tests/unit/general/test_umpire_mem.cpp:112]: The idea of this example was to allocate on the device only and my understanding is that now we're allocating on the host then also on the device, is that correct? Should it just be `Vector dev_perm(num_elems, MemoryType::DEVICE_UMPIRE)`?

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [inline tests/unit/general/test_umpire_mem.cpp:134]: This (the commented out code) is what we'd do in practice. The current set/restore MT approach is actually what we started with but we didn't really like it because it isn't always clear when allocations are happening when someone is using higher level mfem interfaces.

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [inline tests/unit/general/test_umpire_mem.cpp:140]: +1 this propose interface looks good

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [reply tests/unit/general/test_umpire_mem.cpp:177]: That's definitely what we'll do!

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [inline tests/unit/general/test_umpire_mem.cpp:178]: Isn't it true that when a device `MemoryType` is used the memory is only allocated on the device? Given that the `REQUIRE` below passes it seems like it is. Would this change?

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [inline tests/unit/general/test_umpire_mem.cpp:85]: +1 for lazy device allocations.
We may also want to switch the device type after initialization, which was what this example was demonstrating (although that wasn't very clear, I should have documented it), do you have some kind of `UpdateDeviceType` in mind? Alternatively we can add more methods for setting the `MemoryType` in the mfem classes we are using but an update method seems less invasive

File: mesh/mesh.hpp
Comment (v-dobrev) [reply mesh/mesh.hpp:782]: I was thinking of temporarily switching the device memory type: e.g. we can have a macro (or some c++ function with lambdas magic) like this:
```c++
MFEM_WITH_DEVICE_MEMORY(MemoryType::DEVICE_UMPIRE_2,
   auto geom_factors = mesh->GetGeometricFactors(ir, flags);
);
```
We can also add a `MemoryType` option for the device memory but then for the default value we'll need to add another constant in the `MemoryType` enum -- something like `MemoryType::DEVICE_DEFAULT` which will tell the method to use the currently set device memory type.

File: tests/unit/general/test_umpire_mem.cpp
Comment (v-dobrev) [reply tests/unit/general/test_umpire_mem.cpp:112]: The class `Memory` always needs a host pointer (at least in the current implementation) even when you explicitly ask for a specific device memory type. If you want to do pure device allocations, we'll need to think what's the best way to achieve that. However, on most systems host allocations are "allocate on use" (i.e. real memory is used only if you touch it) so always doing the host allocation is probably not a substantial overhead.

File: tests/unit/general/test_umpire_mem.cpp
Comment (v-dobrev) [reply tests/unit/general/test_umpire_mem.cpp:134]: That is a good point, in general.

In this case, the host allocation happens right away, so this should work.

However, as you said, the actual moment of allocation in other cases may be more difficult to track.

File: tests/unit/general/test_umpire_mem.cpp
Comment (v-dobrev) [reply tests/unit/general/test_umpire_mem.cpp:140]: Okay I'll add such a constructor.

File: tests/unit/general/test_umpire_mem.cpp
Comment (v-dobrev) [reply tests/unit/general/test_umpire_mem.cpp:178]: As mentioned in a previous comment, we always perform the host allocation. When the device allocation actually happens (with the constructor used here) is unspecified at the moment. I was thinking of specifying this to be always a "lazy" allocation for the device pointers (currently some device allocations are not "lazy", I think).

File: tests/unit/general/test_umpire_mem.cpp
Comment (v-dobrev) [reply tests/unit/general/test_umpire_mem.cpp:85]: Do you need to use the same vector with two different device memory types? E.g. first you use it with "non-pool" device memory and later you want to move it to "pool" device memory?

File: mesh/mesh.hpp
Comment (tomstitt) [reply mesh/mesh.hpp:782]: This could work! Neat idea!

File: tests/unit/general/test_umpire_mem.cpp
Comment (tomstitt) [reply tests/unit/general/test_umpire_mem.cpp:85]: > first you use it with "non-pool" device memory and later you want to move it to "pool" device memory?

We don't have that use case anywhere.

It may be enough to have a constructor with `h_mt` and `d_mt`, your `MFEM_WITH_DEVICE_MEMORY` idea, and lazy device allocations, I'll look more closely next week

File: general/mem_manager.hpp
Comment (tzanio) [reply general/mem_manager.hpp:44]: Can that be reflected in the name, maybe `DEVICE_UMPIRE_TMP` instead of `DEVICE_UMPIRE_2`?

File: mesh/mesh.hpp
Comment (tzanio) [reply mesh/mesh.hpp:782]: I do like `MFEM_WITH_DEVICE_MEMORY`

**Comment by tzanio** [comment]
Travis is failing because of documentation issues

File: general/mem_manager.hpp
Comment (tomstitt) [reply general/mem_manager.hpp:44]: We used to call it `DEVICE_TEMP_UMPIRE` but any one of these is fine

https://github.com/mfem/mfem/commit/802f21387316fb590dba85ca0e8e8bf8d19da2e4#diff-8c7013577433cbad6586f7299f99abea50089e068ea39a35bdcfb6f8f75f9214R79

Comment (tzanio) [reply general/mem_manager.hpp:44]: `DEVICE_TEMP_UMPIRE` is also good for me

Comment (v-dobrev) [reply general/mem_manager.hpp:44]: The idea is that this type just denotes a second Umpire device allocator -- how it is used is up to the application. That's why I want to avoid calling it `TMP` or `TEMP`.

**Comment by v-dobrev** [comment]
> Travis is failing because of documentation issues

For now, this branch has only the proposed interface changes in the header files (see first post above) -- ignore the CI for now.

**Review by tzanio — APPROVED** [review]

**Comment by tomstitt** [comment]
I got the Umpire test working (`cd test/unit && ./unit_tests UmpireMemorySpace`) but there are still some TODOs and I'm not sure about all of the implementation. I saved the parts I was less sure about for the last commit so it would be easy to revert fi we need

**Comment by v-dobrev** [comment]
This PR is now ready for full review.

File: general/mem_manager.hpp
Comment (v-dobrev) [inline general/mem_manager.hpp:284]: Same here -- was this change in behavior necessary?

File: general/mem_manager.hpp
Comment (v-dobrev) [inline general/mem_manager.hpp:356]: I'll take a look if we can eliminate this restriction.

**Review by artv3 — APPROVED** [review]
Yes! This is high impact work! Will enable large simulations to make more efficient use of device memory!

**Review by tzanio — APPROVED** [review]
Thanks @tomstitt and @v-dobrev !

File: general/mem_manager.hpp
Comment (tomstitt) [reply general/mem_manager.hpp:284]: In the master `New_` has a check like `flags |= IsHostMemory(mt) ? Mem::VALID_HOST : Mem::VALID_DEVICE` but since we now pass `h_mt` and `d_mt` to `New_` and we're doing lazy device allocation I changed it to just be `flags |= VALID_DEVICE`.

I'm thinking we should actually change `if (!IsHostMemory(mt)) { flags |= VALID_DEVICE; }` in the implementation of `Memory::New(size, mt)` to something like: `if (!IsHostMemory(mt)) { flags |= VALID_DEVICE & ~VALID_HOST; }` , which will be the same (I think) as master. How does that sound?

File: general/mem_manager.hpp
Comment (tomstitt) [reply general/mem_manager.hpp:192]: Actually, no, see my other comment

**Comment by v-dobrev** [comment]
Okay, I think I've addressed all discussed changes.

@camierjs, @tomstitt, @artv3, @tzanio, take one more look at this PR and approve if you are okay with it. After approvals, I'll merge it in #1644 for testing.

Before merging this, I'll test it with CUDA and Umpire enabled to make sure the new test, `tests/unit/general/test_umpire_mem.cpp`, works.

**Review by tzanio — APPROVED** [review]

**Comment by v-dobrev** [comment]
With the last commit (4e8b78c) the tests, `make test`, pass when both CUDA and Umpire are enabled.

TODO: update `CHANGELOG`.

**Review by camierjs — APPROVED** [review]
Looks good! Thank you @v-dobrev and @tomstitt for the changes.

**Review by artv3 — APPROVED** [review]

**Review by tomstitt — APPROVED** [review]
I added one additional check to the umpire test to make sure the host allocator isn't being used but otherwise looks great! Thanks @v-dobrev ! I

**Pull request merged**
